src/date.c

```diff
@@ -1,18 +1,984 @@
+    if(ret < 0) { \
+      ret += c2; \
+    } \
+    return ret; \
+  }
+
+DEF_EXCTRACT_TIMEPART(HourFromTime, msPerHour, HoursPerDay)
+DEF_EXCTRACT_TIMEPART(MinFromTime, msPerMinute, MinutesPerHour)
+DEF_EXCTRACT_TIMEPART(SecFromTime, msPerSecond, SecondsPerMinute)
+DEF_EXCTRACT_TIMEPART(msFromTime, 1, msPerSecond)
+
+
+static etimeint_t ecma_MakeTime(etimeint_t hour, etimeint_t min,
+                                    etimeint_t sec, etimeint_t ms) {
+  return ((hour * MinutesPerHour + min)
+                      * SecondsPerMinute + sec) * msPerSecond + ms;
+}
+
+static etimeint_t ecma_MakeDay(int year, int month, int date)
+{
+  int *days;
+
+  etimeint_t yday;
+  etimeint_t mday;
+  
+  year += floor(month / 12);
+  
+  month = month % 12;
+  
+  yday = floor(ecma_TimeFromYear(year) / msPerDay);
+
+  days = ecma_getfirstdays(ecma_IsLeapYear(year));
+
+  mday = days[month];
+  
+  return yday + mday + date - 1;
+}
+
+static etimeint_t ecma_MakeDate(etimeint_t day, etimeint_t time) {
+  return (day * msPerDay + time);
+}
+
+/*++++ ECMA alternatives to struct tm ++++*/
+
+struct timeparts {
+  int year; /* can be negative, up to +-282000 */
+  int month; /* 0-11 */
+  int day; /* 1-31 */
+  int hour; /* 0-23 */
+  int min; /* 0-59 */
+  int sec; /* 0-59 */
+  int msec; 
+  int dayofweek; /* 0-6 */
+};
+
+/*+++ this functions is used to get current date/time & timezone */
+
+static void d_gettime(etime_t *time) {
+  struct timeval tv;
+  gettimeofday(&tv, NULL);
+  *time = (etime_t)tv.tv_sec * 1000 + (etime_t)tv.tv_usec / 1000;
+}
+
+static const char* d_gettzname() {
+  return g_tzname;
+}
+
+static etime_t d_mktime_impl(const struct timeparts *tp) {
+  return ecma_MakeDate(ecma_MakeDay(tp->year, tp->month, tp->day),
+                       ecma_MakeTime(tp->hour, tp->min, tp->sec, tp->msec));
+}
+
+/*++++ libc mktime alternatives ++++*/
+
+static etime_t d_lmktime(const struct timeparts *tp) {
+  return ecma_UTC(d_mktime_impl(tp));
+}
+
+static etime_t d_gmktime(const struct timeparts *tp) {
+  return d_mktime_impl(tp);
+}
+
+typedef etime_t (*fmaketime)(const struct timeparts* );
+
+/*++++ libc gmtime & localtime alternatives ++++*/
+static void d_gmtime(const etime_t *t, struct timeparts *tp) {
+  tp->year = ecma_YearFromTime_s(*t);
+  tp->month = ecma_MonthFromTime(*t, tp->year);
+  tp->day = ecma_DateFromTime(*t, tp->year);
+  tp->hour = ecma_HourFromTime(*t);
+  tp->min = ecma_MinFromTime(*t);
+  tp->sec = ecma_SecFromTime(*t);
+  tp->msec = ecma_msFromTime(*t);
+  tp->dayofweek = ecma_WeekDay(*t);
+}
+
+static void d_localtime(const etime_t* time, struct timeparts *tp) {
+  etime_t local_time = ecma_LocalTime(*time);
+  d_gmtime(&local_time, tp);
+}
+
+typedef void (*fbreaktime)(const etime_t* , struct timeparts* );
+
+/*++++ some isXXX helpers ++++*/
+static val_t d_trytogetobjforstring(struct v7 *v7, val_t obj) {
+  val_t ret = V7_TAG_NAN;
+
+  if(!v7_is_object(obj) || (v7_is_object(obj)
```

Comment:
The naming might be confusing. It basically means "hey interpreter, invoke valueOf() on this value". (more or less, since it handles null and undefined as well)